src/syntax.rs

```diff
@@ -0,0 +1,105 @@
+pub enum DiscreteType {
+    Int,
+    Bool,
+    Float,
+}
+
+pub enum Type {
+    Vector(usize, DiscreteType),
+    Function,
+    Shader,
+}
+
+pub enum ShaderType {
+    Vector,
+    Fragment,
+    Tesselation,
+    // And more!
+}
+
+pub enum BOp {
+    Add,
+    Sub,
+    Mul,
+    Div,
+    Mod,
+    And,
+    Or,
+    Eq,
+    Neq,
+    Gt,
+    Gte,
+    Lt,
+    Lte,
+    BitAnd,
+    BitOr,
+    BitXor,
+}
+
+pub struct Block {
+    pub body: Vec<Box<Expr>>
+}
+pub struct Parameter {
+    name: String,
+    datatype: Type
+}
+
+pub enum AccessType {
+    Name(String),
+    Idx(usize),
+}
+
+pub enum Expr {
+    I(usize),
+    B(bool),
+    F(f32),
+    Vect {
+        is_matrix: bool,
+        datatype: DiscreteType,
+        value: Block,
+    },
+    Def {
+        mutable: bool,
+        name: String,
+        value: Box<Expr>,
+    },
+    Ref(String),
+    App {
+        fname: String,
+        arguments: Block,
+    },
+    BinOp {
+        operator: BOp,
+        e1: Box<Expr>,
+        e2: Box<Expr>,
+    },
+    Compose(String, String),
+    Update{
+        target: String,
+        value: Box<Expr>,
+    },
+    Branch {
+        condition: Box<Expr>,
+        b1: Block,
+        b2: Block,
+    },
+    For {
+        condition: Box<Expr>,
+        variable: Option<String>,
+        body: Block,
+    },
+    Access(String, AccessType),
+    Comment(Vec<String>),
+    Func {
+        name: String,
+        parameters: Vec<Parameter>,
+        return_type: Type,
+        body: Block,
+    },
+    Shader {
```

Comment:
Do we want to have a shader function type? I thought we were thinking about doing shaders as functions directly, and instead introducing types that those functions could use to generate the desired shader effects.

Reply:
Ah I remember now, yes. You're talking about the difference between the functions returning a type, and shaders returning named types?

Reply:
Yeah. A shader is generated through a series of function applications that produce the right types for that shader. Something basic would be Vec4 for position and Vec4 for color, where Position and Color are just type synonyms for those types. We could also be more aggressive about keeping those types separate so we can distinguish the set of operations that can be performed on each, make it a little less ambiguous about whether they can be mixed (or not).

Reply:
OK so I had an idea a bit ago and I will lay it out. We want to make sure our users don't mess up their shaders, right? I was reading a bit of Ada and I like their idea of a specific section for declaration of variables. My idea is to have a specific section at the end of the function for setting variables that are returned! That way we don't have to worry about searching inside ASTs for modifications of the returned variable. This would mean that we don't have a return expression, but I don't think that preventing users from performing an early return is a bad thing.
What do you think?